state/protocol/protocol_state/mutator.go

```diff
@@ -5,57 +5,247 @@ import (
-// No errors are expected during normal operations.
-func (m *Mutator) CommitProtocolState(updater protocol.StateUpdater) func(tx *transaction.Tx) error {
-	return func(tx *transaction.Tx) error {
-		updatedState, updatedStateID, hasChanges := updater.Build()
-		if hasChanges {
-			err := m.protocolStateDB.StoreTx(updatedStateID, updatedState)(tx)
-			if err != nil && !errors.Is(err, storage.ErrAlreadyExists) {
-				return fmt.Errorf("could not store protocol state (%v): %w", updatedStateID, err)
+// handleServiceEvents handles applying state changes which occur as a result
+// of service events being included in a block payload:
+//   - inserting incorporated service events
+//   - updating protocol state for the candidate block
+//
+// Consider a chain where a service event is emitted during execution of block A.
+// Block B contains a receipt for A. Block C contains a seal for block A.
+//
+// A <- .. <- B(RA) <- .. <- C(SA)
+//
+// Service events are included within execution results, which are stored
+// opaquely as part of the block payload in block B. We only validate and insert
+// the typed service event to storage once we process C, the block containing the
+// seal for block A. This is because we rely on the sealing subsystem to validate
+// correctness of the service event before processing it.
+// Consequently, any change to the protocol state introduced by a service event
+// emitted during execution of block A would only become visible when querying
+// C or its descendants.
+//
+// This method will only apply service-event-induced state changes when the
+// input block has the form of block C (ie. contains a seal for a block in
+// which a service event was emitted).
+//
+// All updates that are incorporated in service events are applied to the protocol state by mutating protocol state updater.
+// This method doesn't modify any data from self, all protocol state changes are applied to state updater.
+// All other changes are returned as slice of deferred DB updates.
+//
+// Return values:
+//   - dbUpdates - If the service events are valid, or there are no service events,
+//     this method returns a slice of Badger operations to apply while storing the block.
+//     This includes operations to insert service events for blocks that include them.
+//
+// No errors are expected during normal operation.
+func (m *stateMutator) handleServiceEvents(seals []*flow.Seal) (dbUpdates []func(*transaction.Tx) error, err error) {
+	epochFallbackTriggered, err := m.params.EpochFallbackTriggered()
+	if err != nil {
+		return nil, fmt.Errorf("could not retrieve epoch fallback status: %w", err)
+	}
+
+	parentProtocolState := m.stateMachine.ParentState()
+	epochStatus := parentProtocolState.EpochStatus()
+	activeSetup := parentProtocolState.CurrentEpochSetup
+
+	// never process service events after epoch fallback is triggered
+	if epochStatus.InvalidServiceEventIncorporated || epochFallbackTriggered {
+		return dbUpdates, nil
+	}
```

Comment:
## ⚠️ ❓ 

I am worried this logic is potentially incorrect (specifically not fork-aware) 
https://github.com/onflow/flow-go/blob/f1965e1861dcc2f2e8b23bb60ab36abb758a2635/state/protocol/protocol_state/mutator.go#L111-L114

In the past, we only updated the protocol state with very long time intervals in between updates. Now updates can be in close proximity, potentially differing from fork to fork. In particular the notion of [`InvalidStateTransitionAttempted`](https://github.com/onflow/flow-go/blob/f1965e1861dcc2f2e8b23bb60ab36abb758a2635/model/flow/protocol_state.go#L35) is fork-specific. The `EpochFallbackTriggered`  data base flag is more coarse-grained and  is used to control the Epoch lifecycle across _all forks_. That is fine, the Epoch lifecycle remains to be very slow and only influences a node's action in the future. 

In contrast, the line https://github.com/onflow/flow-go/blob/f1965e1861dcc2f2e8b23bb60ab36abb758a2635/state/protocol/protocol_state/mutator.go#L111-L114 also acts retroactively. If we have the epoch fallback triggered at some later time, the current logic would treat earlier blocks differently, based on future information. This seems dangerous and I would like to avoid this. How we evolve the protocol state should _exclusively_ depend on the ancestry of this fork. 

### Suggestion
* I think we should _solely_ use the [`InvalidStateTransitionAttempted`](https://github.com/onflow/flow-go/blob/f1965e1861dcc2f2e8b23bb60ab36abb758a2635/model/flow/protocol_state.go#L35) flag that is part of the `ProtocolStateEntry` 
* Furthermore, I would suggest that we still process Node Ejection on a best effort basis. Despite Epoch Fallback, we still might need to eject misbehaving nodes

Reply:
Fair enough, the global flag kicks in after finalization so I think it's reasonable that we stop processing forks in such way but I also agree that it's largely redundant and we would like to process ejection events even in fallback mode. 
In this [commit](https://github.com/onflow/flow-go/pull/4868/commits/f57cb5e75f0025417a075de1e8626dbb0f8b7ed9) I have removed dependency on `EpochFallbackTriggered` and also have changed how we check invalid state transition attempted flag.

Reply:
I think this is shaping out very nicely. I still have some concerns regarding the atomicity of the current implementation. I think there won't be a problem the way we are using it right now. Nevertheless I feel there is a very subtle limitation of the algorithm that we currently don't document. I think that is a foot-gun for future extensions, when we introduce other state-changing service events. For illustrative purposes, consider the case where EpochSetup and EpochCommit could occur in the same block. If the EpochCommit fails, the state machine is already in the state including the EpochSetup data. But we drop all data base updates. 

Generally, I would really like to have an algorithm, where we later can add further Service Event types. Multiple of those should be able to occur in the same block. So what happens, when one in the middle fails? do we keep the last interim state, potentially with the need to persist some data in the database? 

I think there is a relatively pragmatic and simple convention, that would also very nicely work in the existing code:
* We introduce the convention that all the service events in one block are applied in one transactional state update by the state machine. Formally, if the state machine leaves the protocol-defined path of normal operation, it ends up in an undefined state. I think that is a good level of abstraction for the [`ProtocolStateMachine`](https://github.com/onflow/flow-go/blob/7ad9e93894051d376ac75172ace4b596fca6b8ae/state/protocol/protocol_state/statemachine.go). In all cases, where the  [`ProtocolStateMachine`](https://github.com/onflow/flow-go/blob/7ad9e93894051d376ac75172ace4b596fca6b8ae/state/protocol/protocol_state/statemachine.go) returns an `InvalidServiceEventError`, it refuses to build. 
* The concept of atomically applying all service events in a block lives in my mind in the `stateMutator`, which also manages the respective db updates and orchestrates the `ProtocolStateMachine`. For me it would be natural have the `stateMutator` also orchestrate the Epoch Fallback Mode:
   * we separate the `ProtocolStateMachine` into two state machines: (a) the happy path state machine (currently the logic in [`stateMachine`](https://github.com/onflow/flow-go/blob/7ad9e93894051d376ac75172ace4b596fca6b8ae/state/protocol/protocol_state/updater.go#L20) and (b) the Epoch Fallback Mode state machine
   * The convention would be as follows:
       * If we are on the happy path, we apply all service events from the block. In largely all cases, this will succeed and no further steps are required by the `stateMutator`. 
       * However, in rare cases of Epoch Fallback, the happy path state machine will encounter an invalid transition while processing the service events in block `B`. In this case, the `stateMutator` will engage with the Epoch-Fallback-Mode state machine (b). It starts the  Epoch-Fallback-Mode state machine with the last known good state, i.e. the parent block's state and applies all service events from block B to the Epoch-Fallback-Mode state machine (b). It will probably ignore all events except for ejection events, but who knows. I think it is valuable to describe the Epoch-Fallback-Mode as a separate state machine 

An alternative could be to change the `stateMutator` API to process one service event at a time. Then we really have a proper notion of "latest valid interim state within a block" in case we enter Epoch fallback mode. The system smart contracts would need to ensure that always a valid state is reached after each service event with the same block ... largely thinking out laude as documentation of my thoughts for the upcoming PR https://github.com/onflow/flow-go/pull/4922

Reply:
Based on your comment I have created a PR which is an attempt of an implementation of dedicated state machine for epoch fallback mode. https://github.com/onflow/flow-go/pull/4931
Would like to discuss it and decide on further steps.

Reply:
> In this [commit](https://github.com/onflow/flow-go/pull/4868/commits/f57cb5e75f0025417a075de1e8626dbb0f8b7ed9) I have removed dependency on EpochFallbackTriggered and also have changed how we check invalid state transition attempted flag.

Suggestion: Given this change (which I agree with), we should change the name of this flag to `InvalidEpochTransitionAttempted`, to be clear that it only affects epoch-related service events.